packages/loopring_v3/contracts/aux/bridge/Bridge.sol

```diff
@@ -0,0 +1,605 @@
+                for (uint i = 0; i < connectorCalls.groups[g].calls.length; i++) {
+                    transfers[i] = InternalBridgeTransfer({
+                        owner: connectorCalls.groups[g].calls[i].owner,
+                        tokenID:  _getTokenID(connectorCalls.groups[g].calls[i].token),
+                        amount: connectorCalls.groups[g].calls[i].amount
+                    });
+                }
+                _storeTransfers(transfers);
+            }
+        }
+    }
+
+    function _executeConnectorCall(ConnectorCalls calldata connectorCalls)
+        external
+    {
+        require(msg.sender == address(this), "UNAUTHORIZED");
+        require(connectorCalls.connector != address(this), "INVALID_CONNECTOR");
+
+        bool trusted = trustedConnectors[connectorCalls.connector];
+
+        if (trusted) {
+            // Execute the logic using a delegate so no extra transfers are needed
+
+            // Do the delegate call
+            bytes memory txData = abi.encodeWithSelector(
+                IBridgeConnector.processCalls.selector,
+                connectorCalls
+            );
+            (bool success, bytes memory returnData) = connectorCalls.connector.delegatecall(txData);
+            if (!success) {
+                assembly { revert(add(returnData, 32), mload(returnData)) }
+            }
+            // TODO: maybe return transfers here to batch all of them together in a single `deposit` call across all trusted connectors.
+        } else {
+            // Transfer funds to the external contract with a real call, so the connector doesn't have to be trusted at all.
+
+            // Transfer funds to the connector contract
+            uint ethValue = 0;
+            for (uint i = 0; i < connectorCalls.tokens.length; i++) {
+                if (connectorCalls.tokens[i].amount > 0) {
+                    if (connectorCalls.tokens[i].token != address(0)) {
+                        connectorCalls.tokens[i].token.safeTransferAndVerify(connectorCalls.connector, connectorCalls.tokens[i].amount);
+                    } else {
+                        ethValue = ethValue.add(connectorCalls.tokens[i].amount);
+                    }
+                }
+            }
+
+            // Do the call
+            IBridgeConnector(connectorCalls.connector).processCalls{value: ethValue, gas: connectorCalls.gasLimit}(connectorCalls);
+        }
+    }
+
+    // Returns the tokenID for the given token address.
+    // Instead of querying the exchange each time, the tokenID
+    // is automatically cached inside this contract to save gas.
+    function _getTokenID(address tokenAddress)
+        internal
+        returns (uint16 cachedTokenID)
+    {
+        if (tokenAddress == address(0)) {
+            cachedTokenID = 0;
+        } else {
+            cachedTokenID = cachedTokenIDs[tokenAddress];
```

Comment:
Maybe, maybe not. Now it definitely does (needs to go through multiple contracts to get the data). With [eip-2929](https://eips.ethereum.org/EIPS/eip-2929) maybe not.

eip-2929 makes it a bit hard to optimize solidity now. It will go live in 4 weeks so we need to optimize for it, but tools (like truffle) don't support the Berlin HF yet, so not easy to test and measure yet.

eip-2929 makes things more logical where reading/calling things the first time is 3x more expensive, but than all other reads/calls are cheap (compared to the behaviour now where each read/call costs the same no matter how many times it was accessed before). So manually batching some stuff may not be necessary any more. But because there are many hidden costs with solidity only measuring the gas costs on the EVM is to be trusted in practice.